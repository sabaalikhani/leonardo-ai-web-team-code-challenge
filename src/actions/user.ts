'use server';

import { redirect } from 'next/navigation';
import { cookies } from 'next/headers';

export type UserData = {
	userName: string;
	jobTitle: string;
};

export const submitUserData = async (state: UserData, data: FormData) => {
	try {
		await new Promise((resolve) => setTimeout(resolve, 1000));

		const userName = data.get('userName');
		const jobTitle = data.get('jobTitle');

		if (!userName || !jobTitle) {
			// TODO: Throw error
		}

		const cookieStore = await cookies();

		// NOTE: add expiration time if required
		cookieStore.set('user_name', userName as string);
		cookieStore.set('job_title', jobTitle as string);
	} catch (error) {
		// TODO: Throw error
	}
	redirect('/');
};
