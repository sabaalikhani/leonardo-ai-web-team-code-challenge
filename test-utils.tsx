import React, { ReactElement } from 'react';
import { render, RenderOptions, screen } from '@testing-library/react';
import { Provider as ChakraProvider } from '@/components/ui/provider';

if (typeof window !== 'undefined') {
	window.matchMedia = jest.fn().mockImplementation((query) => ({
		matches: false,
		media: query,
		onchange: null,
		addListener: jest.fn(),
		removeListener: jest.fn(),
		addEventListener: jest.fn(),
		removeEventListener: jest.fn(),
	}));
}

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
	return <ChakraProvider>{children}</ChakraProvider>;
};

const customRender = (
	ui: ReactElement,
	options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render, screen };
