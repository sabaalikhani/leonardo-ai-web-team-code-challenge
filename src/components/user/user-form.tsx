'use client';

import { useActionState } from 'react';
import { Input, Fieldset } from '@chakra-ui/react';
import { submitUserData } from '@/actions/user';
import type { UserData } from '@/actions/user';
import { Field } from '@/components/ui/field';
import { Button } from '@/components/ui/button';

const initialState = {
	userName: '',
	jobTitle: '',
};

type UserFormProps = {
	data?: UserData;
	testId?: string;
};

export const UserForm = ({
	data = initialState,
	testId = 'user-form',
}: UserFormProps) => {
	const [_, formAction, isPending] = useActionState(submitUserData, data); // eslint-disable-line @typescript-eslint/no-unused-vars 

	return (
		<form action={formAction} data-testid={testId}>
			<Fieldset.Root maxW='md' margin='auto'>
				<Fieldset.Legend>User details</Fieldset.Legend>
				<Fieldset.Content>
					<Field label='User name' required>
						<Input
							name='userName'
							defaultValue={data.userName}
							appearance='primary'
						/>
					</Field>
					<Field label='Job title' required>
						<Input
							name='jobTitle'
							defaultValue={data.jobTitle}
							appearance='primary'
						/>
					</Field>
				</Fieldset.Content>
				<Button
					type='submit'
					alignSelf='flex-start'
					loading={isPending}
					appearance='primary'
					w={['full', 'auto']}
					data-testid='user-form-submit'
				>
					Submit
				</Button>
			</Fieldset.Root>
		</form>
	);
};
