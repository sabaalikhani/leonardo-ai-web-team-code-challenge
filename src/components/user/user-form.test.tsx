import userEvent from '@testing-library/user-event';
import { screen, render } from '@/test-utils';
import { submitUserData } from '@/actions/user';
import { UserForm } from './user-form';

jest.mock('@/actions/user', () => ({
	submitUserData: jest.fn(),
}));

describe('UserForm', () => {
	it('should render the component correctly', () => {
		const { container } = render(<UserForm />);

		expect(screen.getByTestId('user-form')).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it('should render the form with provided data', () => {
		render(<UserForm data={{ userName: 'John', jobTitle: 'Developer' }} />);

		expect(screen.getByDisplayValue('John')).toBeInTheDocument();
		expect(screen.getByDisplayValue('Developer')).toBeInTheDocument();
	});

	it('should submit the form upon clicking the submit button', async () => {
		const user = userEvent.setup();

		render(<UserForm />);

		await user.type(
			screen.getByRole('textbox', { name: 'User name' }),
			'John'
		);
		await user.type(
			screen.getByRole('textbox', { name: 'Job title' }),
			'Developer'
		);

		await user.click(screen.getByRole('button', { name: 'Submit' }));

		expect(submitUserData).toHaveBeenCalledTimes(1);
	});
});
