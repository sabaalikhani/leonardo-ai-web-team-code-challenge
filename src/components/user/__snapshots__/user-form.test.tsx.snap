// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm should render the component correctly 1`] = `
<div>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
    data-testid="user-form"
  >
    <fieldset
      class="fieldset__root css-11gbs6y"
      data-part="root"
      data-scope="fieldset"
    >
      <legend
        class="fieldset__legend css-1ivdo4g"
        data-part="legend"
        data-scope="fieldset"
      >
        User details
      </legend>
      <div
        class="fieldset__content css-mpjbgj"
      >
        <div
          class="chakra-field__root css-13f87cb"
          data-part="root"
          data-scope="field"
          id="field:::r1:"
          role="group"
        >
          <label
            class="chakra-field__label css-65zcwe"
            data-part="label"
            data-scope="field"
            for=":r1:"
            id="field:::r1:::label"
          >
            User name
            <span
              aria-hidden="true"
              class="chakra-field__requiredIndicator css-gp8wgo"
            >
              *
            </span>
          </label>
          <input
            class="chakra-input css-b3diqb"
            data-part="input"
            data-required=""
            data-scope="field"
            id=":r1:"
            name="userName"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-field__root css-13f87cb"
          data-part="root"
          data-scope="field"
          id="field:::r2:"
          role="group"
        >
          <label
            class="chakra-field__label css-65zcwe"
            data-part="label"
            data-scope="field"
            for=":r2:"
            id="field:::r2:::label"
          >
            Job title
            <span
              aria-hidden="true"
              class="chakra-field__requiredIndicator css-gp8wgo"
            >
              *
            </span>
          </label>
          <input
            class="chakra-input css-b3diqb"
            data-part="input"
            data-required=""
            data-scope="field"
            id=":r2:"
            name="jobTitle"
            required=""
            value=""
          />
        </div>
      </div>
      <button
        class="chakra-button css-1hceb3j"
        data-testid="user-form-submit"
        type="submit"
      >
        Submit
      </button>
    </fieldset>
  </form>
</div>
`;
